src/query/graphite/native/builtin_functions.go

```diff
@@ -1613,64 +1662,88 @@ func movingMedian(ctx *common.Context, _ singlePathSpec, windowSize string) (*bi
-			if windowPoints <= 0 {
-				err := errors.NewInvalidParamsError(fmt.Errorf(
-					"non positive window points, windowSize=%s, stepSize=%d",
-					windowSize, series.MillisPerStep()))
+	return &binaryContextShifter{
+		ContextShiftFunc:  contextShiftingFn,
+		BinaryTransformer: func(bootstrapped, original ts.SeriesList) (ts.SeriesList, error) {
+			bootstrapList, err := combineBootstrapWithOriginal(ctx,
+				bootstrapStartTime, bootstrapEndTime,
+				bootstrapped, singlePathSpec(original))
+			if err != nil {
 				return ts.NewSeriesList(), err
 			}
-			window := make([]float64, windowPoints)
-			util.Memset(window, math.NaN())
-			numSteps := series.Len()
-			offset := bootstrap.Len() - numSteps
-			vals := ts.NewValues(ctx, series.MillisPerStep(), numSteps)
-			for i := 0; i < numSteps; i++ {
-				for j := i + offset - windowPoints; j < i+offset; j++ {
-					if j < 0 || j >= bootstrap.Len() {
-						continue
-					}
-
-					idx := j - i - offset + windowPoints
-					if idx < 0 || idx > len(window)-1 {
-						continue
-					}
 
-					window[idx] = bootstrap.ValueAt(j)
+			results := make([]*ts.Series, 0, original.Len())
+			maxWindowPoints := 0
+			for i, _ := range bootstrapList.Values {
+				series := original.Values[i]
+				windowPoints := windowPointsLength(series, interval)
+				if windowPoints <= 0 {
+					err := errors.NewInvalidParamsError(fmt.Errorf(
+						"non positive window points, windowSize=%s, stepSize=%d",
+						windowSize.stringValue, series.MillisPerStep()))
+					return ts.NewSeriesList(), err
 				}
-				nans := common.SafeSort(window)
-				if nans < windowPoints {
-					index := (windowPoints - nans) / 2
-					median := window[nans+index]
-					vals.SetValueAt(i, median)
+				if windowPoints > maxWindowPoints {
+					maxWindowPoints = windowPoints
 				}
 			}
-			name := fmt.Sprintf("movingMedian(%s,%q)", series.Name(), windowSize)
-			newSeries := ts.NewSeries(ctx, name, series.StartTime(), vals)
-			results = append(results, newSeries)
-		}
 
-		original.Values = results
-		return original, nil
-	}
+			windowPoints := make([]float64, maxWindowPoints)
+			for i, bootstrap := range bootstrapList.Values {
+				series := original.Values[i]
+				currWindowPoints := windowPointsLength(series, interval)
+				window := windowPoints[:currWindowPoints]
+				util.Memset(window, math.NaN())
+				numSteps := series.Len()
+				offset := bootstrap.Len() - numSteps
+				vals := ts.NewValues(ctx, series.MillisPerStep(), numSteps)
+				for i := 0; i < numSteps; i++ {
+					for j := i + offset - currWindowPoints; j < i+offset; j++ {
+						if j < 0 || j >= bootstrap.Len() {
+							continue
+						}
 
-	return &binaryContextShifter{
-		ContextShiftFunc:  contextShiftingFn,
-		BinaryTransformer: transformerFn,
+						idx := j - i - offset + currWindowPoints
+						if idx < 0 || idx > len(window)-1 {
+							continue
+						}
+
+						window[idx] = bootstrap.ValueAt(j)
+					}
+					impl(window, &vals, currWindowPoints, i)
```

Comment:
After changing *ts.MutableValues to just ts.MutableValues this should just be `vals` instead of `&vals`.